src/cargo/util/toml/mod.rs

```diff
@@ -2073,6 +2122,22 @@ impl TomlManifest {
     }
 }
 
+fn inheritable_from_path(
+    config: &Config,
+    resolved_path: PathBuf,
+) -> CargoResult<InheritableFields> {
+    let key = resolved_path.parent().unwrap();
+    let source_id = SourceId::for_path(key)?;
+    let (man, _) = read_manifest(&resolved_path, source_id, config)?;
+    match man.workspace_config() {
+        WorkspaceConfig::Root(root) => Ok(root.inheritable().clone()),
+        _ => bail!(
+            "root of a workspace inferred but wasn't a root: {}",
```

Comment:
Just to double check, this is mostly an internal error because we'll only call this on verified inputs?

Reply:
If a `Cargo.toml` specified `workspace = ".."` but that manifest was not a workspace root, it would throw this error. It would have to take the path below when searching for a workspace root.
```rust
WorkspaceConfig::Member { root: Some(ref path_to_root) } => {
    debug!("find_root - found pointer");
    Ok(Some(read_root_pointer(&ances_manifest_path, path_to_root)))
}
```

Reply:
Yes, I see the test case now (should have looked for it first).  I feel like we should improve the error to make that clearer.  Here, we have several branches that get to this error, with all but one being internal.  We can either tailor this message to that one branch or we can keep make this all internal and updated `read_root_pointer` to return an error if the pointer is bad.  Putting the error in `read_root_pointer` would also allow us to tell users what manifest had the bad pointer.  Since this isn't already done, I'm unsure what repercussions this might have.

Reply:
I pulled the error message from 
```rust
pub fn load_workspace_config(&mut self) -> CargoResult<Option<WorkspaceRootConfig>> {
    // If we didn't find a root, it must mean there is no [workspace] section, and thus no
    // metadata.
    if let Some(root_path) = &self.root_manifest {
        let root_package = self.packages.load(root_path)?;
        match root_package.workspace_config() {
            WorkspaceConfig::Root(ref root_config) => {
                return Ok(Some(root_config.clone()));
            }

            _ => bail!(
                "root of a workspace inferred but wasn't a root: {}",
                root_path.display()
            ),
        }
    }

    Ok(None)
}
```

It looks like we could just pull it out and make `read_root_pointer` check if it's a `WorkspaceConfig::Root` but that might get complicated since one uses `.load()` and the other juts parses the manifest and doesn't save it.

Reply:
Oh, this is existing?  Let's just leave it the same